quic/s2n-quic-core/src/datagram/default.rs

```diff
@@ -78,10 +232,178 @@ impl Builder {
+    // are there
+    let second = default_sender.queue.pop_front().unwrap();
+    assert_eq!(second.data[..], [4, 5, 6]);
+    let third = default_sender.queue.pop_front().unwrap();
+    assert_eq!(third.data[..], [7, 8, 9]);
+    assert!(default_sender.queue.is_empty());
+}
+
+#[test]
+fn send_datagram_with_error() {
+    let conn_info = ConnectionInfo {
+        max_datagram_payload: 100,
+    };
+    // Create a default sender queue that only holds two elements
+    let mut default_sender = DefaultSender::builder()
+        .with_capacity(2)
+        .with_connection_info(conn_info)
+        .build()
+        .unwrap();
+    let datagram_0 = bytes::Bytes::from_static(&[1, 2, 3]);
+    let datagram_1 = bytes::Bytes::from_static(&[4, 5, 6]);
+    let datagram_2 = bytes::Bytes::from_static(&[7, 8, 9]);
+    assert_eq!(default_sender.send_datagram(datagram_0), Ok(None));
+    assert_eq!(default_sender.send_datagram(datagram_1), Ok(None));
+    // Attempting to send a third datagram will result in an error, since the queue
+    // is at capacity
+    assert_eq!(
+        default_sender.send_datagram_with_error(datagram_2),
+        Err(SendDatagramError::QueueAtCapacity)
+    );
+
+    // Check that the first two datagrams are still there
+    let first = default_sender.queue.pop_front().unwrap();
+    assert_eq!(first.data[..], [1, 2, 3]);
+    let second = default_sender.queue.pop_front().unwrap();
+    assert_eq!(second.data[..], [4, 5, 6]);
+    assert!(default_sender.queue.is_empty());
+}
+
+#[test]
+fn poll_send_datagram() {
+    use futures_test::task::new_count_waker;
+
+    let conn_info = ConnectionInfo {
+        max_datagram_payload: 100,
+    };
+    // Create a default sender queue that only holds two elements
+    let mut default_sender = DefaultSender::builder()
+        .with_capacity(2)
+        .with_connection_info(conn_info)
+        .build()
+        .unwrap();
+    let datagram_0 = bytes::Bytes::from_static(&[1, 2, 3]);
+    let datagram_1 = bytes::Bytes::from_static(&[4, 5, 6]);
+    let datagram_2 = bytes::Bytes::from_static(&[7, 8, 9]);
+
+    let (waker, _counter) = new_count_waker();
+    let mut cx = Context::from_waker(&waker);
+
+    assert_eq!(
+        default_sender.poll_send_datagram(datagram_0, &mut cx),
+        Poll::Ready(Ok(()))
+    );
+
+    assert_eq!(
+        default_sender.poll_send_datagram(datagram_1, &mut cx),
+        Poll::Ready(Ok(()))
+    );
+
+    // Waker has not been set up yet because queue is not yet at capacity
+    assert!(default_sender.waker.is_none());
+    assert_eq!(
+        default_sender.poll_send_datagram(datagram_2, &mut cx),
+        Poll::Pending
+    );
+
+    // Since queue is at capacity default_sender is now storing a waker that will
+    // alert when the queue has more space
+    assert!(default_sender.waker.is_some());
+
+    // Check that the first two datagrams are on the queue
+    let first = default_sender.queue.pop_front().unwrap();
```

Comment:
ideally we'd call the `on_transmit` function here, which would call `waker.wake` and we could assert that we're getting woken up on new capacity.